src/action/list_directory.rs

```diff
@@ -0,0 +1,533 @@
+                offset: None,
+                recursion_depth: None,
+            }),
+            iterator: None,
+        }
+    }
+
+    #[test]
+    fn test_empty_pathspec_field() {
+        let request: Result<super::Request, _> =
+            Request::from_proto(ListDirRequest {
+                pathspec: None,
+                iterator: None,
+            });
+        assert!(request.is_err());
+    }
+
+    #[test]
+    fn test_empty_path_options() {
+        let request: Result<super::Request, _> = Request::from_proto
+            (fill_proto_request(None, Some(0), Some(String::from("/"))));
+        assert!(request.is_ok());
+    }
+
+    #[test]
+    fn test_unset_pathtype() {
+        let request: Result<super::Request, _> = Request::from_proto
+            (fill_proto_request(None, Some(-1), Some(String::from("/"))));
+        assert!(request.is_err());
+    }
+
+
+    #[test]
+    fn test_empty_pathtype() {
+        let request: Result<super::Request, _> = Request::from_proto
+            (fill_proto_request(None, None, Some(String::from("/"))));
+        assert!(request.is_err());
+    }
+
+    #[test]
+    fn test_empty_path() {
+        let request: Result<super::Request, _> = Request::from_proto
+            (fill_proto_request(None, Some(0), None));
+        assert!(&request.is_ok());
+        assert_eq!(request.unwrap().pathspec.path, PathBuf::from("/"));
+    }
+
+    #[test]
+    fn test_empty_dir() {
+        let dir = tempdir().unwrap();
+        let request = super::Request {
+            pathspec: PathSpec {
+                path_options: None,
+                pathtype: PathType::OS,
+                path: PathBuf::from(dir.path()),
+            }
+        };
+        let mut session = session::test::Fake::new();
+        assert!(handle(&mut session, request).is_ok());
+        assert_eq!(session.reply_count(), 0);
+    }
+
+    #[test]
+    fn test_nonexistent_path() {
+        let dir = tempdir().unwrap();
+        let request = super::Request {
+            pathspec: PathSpec {
+                path_options: None,
+                pathtype: PathType::OS,
+                path: PathBuf::from(dir.path().join("nonexistent_subdir")),
+            }
+        };
+        let mut session = session::test::Fake::new();
+        assert!(handle(&mut session, request).is_err());
+    }
+
+    #[test]
+    fn test_lexicographical_order() {
+        let dir = tempdir().unwrap();
+        let dir_path = dir.path();
+        let file_names = vec!["1numfile", "5numfile",
+                              "Afile", "aafile", "afile",
+                              "bfile", "zfile", "*file", "(file)", "=^-^="];
+        let mut file_paths = vec![];
+        for file_name in &file_names {
+            file_paths.push(dir_path.join(file_name));
+        }
+        file_paths.shuffle(&mut thread_rng());
+        for file_path in &file_paths {
+            std::fs::File::create(file_path).unwrap();
+        }
```

Comment:
Instead of this complicated logic, try to keep your tests simple. Here, I think it would be more readable to have explicit `std::fs::File::create` calls with explicit paths, instead of using `for`s and temporary arrays, e.g.:

```rust
std::fs::File::create(dir_path.join("foo")).unwrap();
std::fs::File::create(dir_path.join("bar")).unwrap();
```

I can see some value in having a different order each time you execute the tests, but I don't think the cost (unreadable test code) is worth it. It also introduces a problem, that (unlikely, but possible) your `shuffle` call with give you an already ordered input and the test will pass even if your action does not sort the results.

Reply:
Fixed.